Add architecture files (#6)
hassanobeid1994/tr_b_causal_2020: notebooks/_06-Architecture.md

    display_name: Python 3
    language: python
    name: python3
---

# Architecture

## 1. At a “high” level, how will the project deliver each of the needed requirements?
- **Try to think in terms of some manageable (e.g. 10-20) steps**
- **List needed intermediate products (e.g. pseudo-code, unit tests, etc.)**
- **Try to be concise but comprehensive?**

The project is split into two tasks:

The first task will be based on simulation work and will go as follows:

For a causal graph where all Xs are independent:
 - Ingest data
 - Manipulate data as needed, the output of this step will be a cleaned dataset
 - Specify a choice model based on the data at hand, the specification should follow required conventions by third party dependencies
 - The cleaned dataset will be used in a function that simulates choice data based on the specified model, resulting in a dataset with simulated choices
 - The dataset will be used to estimate a choice model and recover the necessary parameters. These parameters can be stored in a dictionary.
 - The previous two steps will be repeated for N times (say 100) and the parameters from each iteration will be stored in a dictionary entry.

>  - A function that takes in the parameters of the estimated model and results in the percentage of time that the true parameter is recovered.

Reviewer: How will we deal with the fact that the parameter vector is multivariate?

Will we report parameter recovery percentages separately for each parameter?

Will we have some way of aggregating results over all parameters and then reporting only a single percentage?

Or will we report parameter recovery percentages for a subset of parameters? (perhaps just the parameter(s) we expect to not be recovered well?).

Author: I think for most flexibility, we can make sure we get recovery percentages for individual, a subset, and all parameters. However, I think it makes sense to start with individual parameters then build on that. Let me know what you think.

Reviewer: That sounds good to me--track each parameter individually, and we can do whatever afterwards.

An easy way to do this might be something along the lines of 
```
simulated_parameters = np.empty((NUM_SIMULATIONS, num_choice_model_parameters), dtype=float)
for i in range(NUM_SIMULATIONS):
    current_data = simulate_data()
    simulated_parameters[i, :] = estimate_parameters(current_data)
np.savetxt(SIM_EXPERIMENT_RESULT_PATH, simulated_paramters, fmt='%.18f')
```

Author: ah I see, if you look at the code I have now, all it does is stores each of the fitted model objects from pylogit as a dictionary value. We can access every single instance of that object by calling it individually. But I like this as well!!

Reviewer: Yes what you did makes sense and completely works!

For a realistic causal graph with confounding:
 - Ingest data
 - Manipulate data as needed, the output of this step will be a cleaned dataset
 - Specify a choice model based on the data at hand, the specification should follow required conventions by third party dependencies
 - Specify the realistic causal graph with desired confounding structures. The causal graph will be saved as a causalgraphicalmodels object and can be saved as .png file
 - The cleaned dataset will be used in a function that simulates choice data based on the specified model, resulting in a dataset with simulated choices
 - The dataset will be used to estimate a choice model and recover the necessary parameters. These parameters can be stored in a dictionary.
 - The previous two steps will be repeated for N times (say 100) and the parameters from each iteration will be stored in a dictionary entry.

>  - A function that takes in the parameters of the estimated model and results in the percentage of time that the true parameter is recovered.

Reviewer: See question above about how to deal with multivariate parameter vectors. How will we report the results given that we have more than one parameter being estimated?

Author: I think for most flexibility, we can make sure we get recovery percentages for individual, a subset, and all parameters. However, I think it makes sense to start with individual parameters then build on that. Let me know what you think.

Reviewer: See my reply above:

Your suggestion sounds good to me!---track each parameter individually, and we can do whatever afterwards.

An easy way to do this might be something along the lines of 
```
simulated_parameters = np.empty((NUM_SIMULATIONS, num_choice_model_parameters), dtype=float)
for i in range(NUM_SIMULATIONS):
    current_data = simulate_data()
    simulated_parameters[i, :] = estimate_parameters(current_data)
np.savetxt(SIM_EXPERIMENT_RESULT_PATH, simulated_paramters, fmt='%.18f')
```

## 2. Description of major components of the system and their relations:
### 2.1 What should each component do?
Simulation work:
Simulation functions will generate data to be used in modeling efforts
Choice model estimation will try to recover the true parameters of the specified model

### 2.2 What is the interface between each component and each other component?

> Each component(classes) will be written to be as independent as possible from other components. Functions within classes

Reviewer: In Q2.1, you mention that a simulation object will output data, and that an estimation object will take in that data. This is the interface for the two components (as in the point where the two classes meet) between the two classes. Sorry if it wasn't clear earlier!

Reviewer: Also, note there should be some sort of function or method that makes ssure that the data that the simulation object outputs is in some guaranteed form that the estimation object is expecting. For instance, do we have all of the correct columns? Is the data in the correct format (long vs wide format), etc.

Author: Perfect.

 from question 4, calls on the functions in each of the classes to simulate data, estimate models, and provide
 parameters and statistics relevant to the problem at hand.

## 6. Description of algorithms:
### 6.1  What, if any, computational algorithms are being implemented as part of this project instead of relying upon external implementations?
Hassan to fill in based on the deconfounder work.

### 6.2 What purpose(s) do those algorithms serve?
See above

### 6.3 Are each of these algorithms fully understood by someone on the project team?
See above

### 6.4  What are alternative algorithms that could have served the purposes of the algorithms we are implementing?
See above

### 6.5 Why are we using the particular algorithms we’ve chosen instead of others? 
See above

### 6.6 Why do we believe the algorithms can be implemented?
See above

### 6.7 Do we have any concerns with regard to algorithmic efficiency?
See above

## 7. Description of data usage in the system:
### 7.1 How will data be ingested / accessed?
Data will be ingested using pandas from either an online path to the dataset or from a local path on the user's machine

### 7.2 How will data be validated?

> Data can be validated using third party dependencies such

Reviewer: So far, the only data being used is the data from my asymmetric paper, so the only validation that comes to mind is just that we reproduce the model results in `references/code/_50_tb-cordon_toll_general_models--simplify_the_models.ipynb`. That provides confidence of no data changes since the asymmetric paper.

Author: Makes sense.

### 7.3 How will data be organized?
Data will be organized in tabular format (.csv)

### 7.4 How will data be transformed?
Data will be transformed using methods available within Pandas

### 7.5 How will created data be output and stored?
Causal graphs will be output to .png files and pushed to the repo

> Summaries of model estimate will be saved as markdown tables

Reviewer: Hmm, maybe for presentation to others.

For ourselves, I would suggest storing model estimates in some easily-read-by-python data format such as a json / pickle / csv file or anything like those. We should be able to easily re-load any previously estimated model parameters or important summary information.

Author: For now, if you look at the simulation, I stored all the fitted models from the simulation in a dictionary, which we can pickle and load again. Let me know if you think dictionaries are good for that.

Plots of results will be saved as .png files and pushed to the repo

### 7.6 How will created data be validated?

> There are packages that will validate data.

Reviewer: As far I can tell, there are two sources of simulated data, based on:
- fitted distributions to observational data
- manually-specified distributions for each node in a causal graph

We should ensure that the simulated data is as we expect.
- Do we have all the columns we expect?
- Do all columns have the correct types?
- Do numeric columns have expected ranges?
- Do numeric columns have means and variances that are within our expectations?
- Do we have the correct number of simulated observations?

Author: Ah got it. Will update my response accordingly.

### 7.7 How does all of the above differ online vs offline? (e.g. we may throw a critical error offline to prevent training with bad data but only log a warning online to note that the data in a request was unexpected.)
N/A

### 7.8 How will the flow / manipulation of data be controlled (and recorded / version-controlled) and made reproducible? E.g. data ingestion and manipulation to create training / testing sets.
- **Flyte?**
- **Make?**
- **Other?**  

> We don't have any plans for this. please make any suggestions as you see needed/fit.

Reviewer: See issue #25 for my suggestions!

Author: Done!

## 8. Description of experimentation protocols (if relevant):

> **If experiments need to be run (e.g. hyper-parameter tuning, model-selection, A/B tests, etc.):**

Reviewer: In this project, the experiments to be run are:
1. Changing a pre-specified causal graph from using independent X's to X's that cause other X's and comparing the change in the percentage of simulations where the data-generating parameters are recovered.
2. Changing one's estimation method from an MNL that doesn't account for unobserved confounding to an MNL model that does account for unobserved confounding.

In both cases, the general experiment may be re-run with different hyper-parameters. Imagine using two different distributions for the latent confounder that is excluded from one's graph. Or imagine using different causal graphs for the first case where some X's cause other X's.

In both scenarios, we may actually have to re-run the experiment multiple times.
- Joan may ask us to change the causal graph we're using.
- We might not get the results we're looking for in the first simulation and need to try differing distributions to make our point.
- Later reviewers may dislike our chosen simulations and ask us to perform other ones (has happened to me in the past)

For all these reasons, we should set up an organized & reproducible way to run and track the experiments.

Author: Sounds good. Adding additional information immediately. let me know what you think.

#### 10.3.1 How does the system decide how much of an external resource is needed? (e.g. how many virtual CPUs are required during training?) 
N/A

## 11. Description of how the system will scale:
### 11.1 With increasing dataset sizes: 
There is no current plan for scaling the system to adapt to large datasets.

### 11.2 With increasing numbers of models being used:
The number of models used should not affect the scalability of the project.

### 11.3 With increasing numbers of parameters:
N/A

### 11.4 With increasing numbers of divergences/losses:
N/A

## 12. Description of how the system will interface with external systems:
- **E.g. if a model needs to be served in Go but it was trained in Python…**
- **If the system needs to manipulate Kubernetes for training...**

N/A

## 13. Description of how errors will be handled:
### 13.1 What are common expected errors from users and how can we guard against them?
Having a wrong specification for a desired model, pylogit has built-in mechanisms to guard against it
Having a wrong specification for a causal graph, we can guard fro

### 13.2 Will we try to fix errors or merely notify users of the error’s presence?

> We will only notify users of the error's presence

Reviewer: ```suggestion
We will only notify users of the error's presence.
```

### 13.3 When errors are encountered, will we quit immediately or wait until some specified point before notifying users of errors? For which errors is each strategy appropriate?
Quit immediately in the case of all errors.

### 13.4 What are the conventions for error messages that the system reports?

### 13.5 Where are errors processed? At the point of detection, by a central error handling class, by functions above in the call stack, etc.
Errors are processed at the point of detection.

### 13.6 What level of responsibility does each class have in validating its own input? Is there a central class (or set of classes) that performs all validation? When can classes assume clean information?
Code for input validation can be built in separate functions and be used part of analysis functions as seen needed.

### 13.7 How will we use (or not use) type hinting to help prevent errors?

> Docstrings will suggest to the users what each of the function takes in as types and what it returns.

Reviewer: ```suggestion
Docstrings will suggest to the users what each of the functions takes in as types and what it returns.
```

Reviewer: Also, I would suggest adding explicit [mypy](http://mypy-lang.org/) typing hints to our docstrings.

I found mypy to be incredibly helpful in past projects for catching type errors that may have gone unnoticed due to lack of sufficiently stringent tests or lack of input validation methods.

Author: Thanks for the suggestion, Tim. Definitely including it.

### 14.3 Are there any parts of the architecture that we do not know how to test? How will these knowledge gaps be resolved before the coding begins?
Given our preliminary knowledge on testing, there will be elements that we won't know how to test. 
Reading through available documentation and tutorials for pytest or other packages will help resolve any knowledge gaps 
in testing any of the parts of the architecture.

### 14.4 How will discovered bugs be turned into test cases?
The guide provided by timothy on how to debug code will be followed to move from finding the bug to testing it.

### 14.5 How will the software be reviewed in general (beyond automated tests) to reduce the probability of programming bugs into the code / increase the chance of producing correctly functioning software?
cross-review between team members will help in detecting any bugs.

## 15. Description of project-sharing procedures:
### 15.1 Is there a plan in place for producing documentation for the software?
- **Documentation for operating the software**
- **Documentation of software limits**
- **Usage examples and tutorials**
- **Key project learnings / retrospective notes**

> Documentation, in the form of Jupyter Notebook accompanied by a Markdown file (README.md) detailing the procedure for installing the environment and the content of folders and notebooks. Any limitations to the project will be noted in the jupyter notebooks developed to illustrate the analysis effort. The README.md markdown file will also will point to documentation pages for any packages used in the project.

Reviewer: In addition to what you've written, we should DEFINITELY include docstrings for all classes and public (i.e. non-underscored) functions used in the project.

A user should be able to call help on any function or class method we've written and learn all they need to know to use the function as intended.

Author: Agreed.

## 17. Description of project tools:
### 17.1 What are all the tools and major external systems upon which the project depends?
Standard scientific packages such as: numpy, pandas, scipy, pylogit, causalgraphicalmodels(or others),
seaborn/matplotlib, scikitlearn, pyro, fitter, pytest. This list is not exhaustive and could grow.

### 17.2 Is there evidence that we understand the basics of how all of those tools work?
- **Have we read any/all relevant documentation and instruction materials?**
- **Have we taken any basic training courses in any of these tools?**

Each team members is individually familiar working with the majority of these packages. Collectively, the team is
familiar working with all of these packages. If additional packages are needed and no team member has the expertise
needed to work with them, the team will make sure to read any available documentation or use examples available. 

## 18. Description of project risks and mitigation efforts:
### 18.1  There should be some evidence for, and of, project feasibility:
- **How do we know that every part of the architecture will work?**
- **Based on the architecture, what are all needed resources?**

> The architecture is mainly based on third party dependencies that have proven to work. The needed resources will be

Reviewer: The first part of the question is "How do we know that every part of the architecture will work?"

Consider Q1. of the architecture doc. There, two main steps are listed for the project.

How do you know that each of these two steps will work? More specifically,
  - What precise papers support the belief that they'll work?
  - What prototype work (feel free to link pull requests / commits) supports the belief that the tasks will work?
  - What logical arguments support the belief that these tasks will work? 
     - If there is no evidence besides the logic, what is the smallest prototype that can be done to provide some evidence in support of the logic

Similar "how do you know it will work" questions apply to Q2 and Q3 of the architecture doc.

Author: I will be addressing this in detail after our meeting.

Author: @timothyb0912 I think this will need to be revisited based on discussions from last week and from our weekly mtg today

one's laptop with an installed environments/requirements and compatible python version.

### 18.2 What could render the project infeasible?
- **Is there adequate resourcing?**

We do not envision inadequare resourcing problems to occur.

### 18.3 Are there known sources of difficulty based on the planned tasks?
The sources of difficulty are related to learning proper software engineering practices and learning about new
third party dependencies.

### 18.4 Are there mismatches between prior knowledge and required tasks?
Yes, this is mitigated by reading literature on needed methodology or documentation of any third party dependencies.

### 18.5 What are the weakest areas of the architecture?
- **Where is the project most vulnerable to failure?**
- **How are these weaknesses being addressed?**

> At this moment, the weakest areas of the architecture are the lack of tests in the available code and the lack of 

Reviewer: I would add two things.

1. The lack of a project workflow file to enable reproduction of the final results. I think this one will be fairly straightforward to make, assuming we follow good practices of creating modular notebooks and code.
2. The lack of a reproducible experiment system. We have no way to track all the different ideas that are likely about to be tried in the next month to get final results. Afterwards, we will then be unable/less-able to reproduce our work along the way to the final results. As a result, we'll be unable/less-able to harvest any useful information from this period since it's more difficult to use the knowledge of what didn't work if you don't remember all the things that didn't work.

Author: Thanks!

 - **clear and unambiguous**
 - **capable of dealing with stochastic functions / objects**
 - **minimizing reliance / use of stochasticity**
 - **fast to evaluate**

This will be filled in as we learn more about tests. The first thing that comes to mind is running the same tests
many times and then verify that the output statistics match one's expectations. In order for tests to be fast, each
test will need to focus on one functionality. We would need to use descriptive names to understand what the tests are
doing. Each test should also be independent from other tests.

### 14.2 What are the various types of tests that are needed?
Unit tests/Feature tests, integration tests.

### 14.3 Are there any parts of the architecture that we do not know how to test? How will these knowledge gaps be resolved before the coding begins?
Given our preliminary knowledge on testing, there will be elements that we won't know how to test. 
Reading through available documentation and tutorials for pytest or other packages will help resolve any knowledge gaps 
in testing any of the parts of the architecture.

### 14.4 How will discovered bugs be turned into test cases?

> The guide provided by timothy on how to debug code will be followed to move from finding the bug to testing it.

Reviewer: ```suggestion
The guide provided by Timothy on how to debug code will be followed to move from finding the bug to testing it.
```

### 14.5 How will the software be reviewed in general (beyond automated tests) to reduce the probability of programming bugs into the code / increase the chance of producing correctly functioning software?

> cross-review between team members will help in detecting any bugs.

Reviewer: ```suggestion
Cross-review between team members will help in detecting any bugs.
```

  jupytext:
    formats: ipynb,md
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.2'
      jupytext_version: 1.4.1
  kernelspec:
    display_name: Python 3
    language: python
    name: python3
---

# Architecture

## 1. At a “high” level, how will the project deliver each of the needed requirements?
- **Try to think in terms of some manageable (e.g. 10-20) steps**
- **List needed intermediate products (e.g. pseudo-code, unit tests, etc.)**
- **Try to be concise but comprehensive?**

The project is split into three main tasks:

1. The first task is focused on a selection-on-observables problem:

For a causal graph with independent covariates:
 1. Ingest data
 2. Manipulate data as needed, the output of this step will be a cleaned dataset
 3. Specify a choice model based on the data at hand, the specification should follow required conventions by third party dependencies
 4. The cleaned dataset will be used in a function that estimated the distributions for variables of interest, this function will produce a parameter dictionary containing the type of variables, the best fitted distribution, and that distribution parameters.

>  5. The resulting parameter dictionary from the previous step will be used to simulate needed data

Reviewer 2: For independent variables only. I didn't see anywhere here that you'll be using the causal graph structure for simulating variables that have parents

 9. Based on observed availability of allternatives, a function will use the cleaned dataset to simulate an availability matrix for alternatives for the simulated dataset
 10. The specified model from step 3 will be used to simulate choices for the simulated dataset.
 11. The dataset will be used to estimate a choice model and recover the necessary parameters. These parameters can be stored in a dictionary.
 12. Steps 7-11 will be repeated for N times (say 100) and the parameters from each iteration will be stored in a dictionary entry.
 13. These paramaters will be stored in a dictionary that will be used in a function to plot the original parameters used to simulate the choices (from step 3) along with any recovered parameters from step 9.
 14. One variable will be perturbed to simulate a potential external intervention. In this scenario, descendant children nodes of the perturbed node will also be perturbed following the estimated relationships from step 8.
 15. Causal effects will be estimated using the models estimated in step 3 and 121.

>  16. Plots will be produced to show the distribution of "estimated" and "true" causal effects

Reviewer 2: I see what you're doing now. Here are my high level comments: 

- The exercise of causal graph with independent variables only is not necessary and does not contribute to our story. All we're doing there is estimate a model on a simulated dataset and seeing if we recover the parameters of the simulation, which is trivial. That used to be our plan but not anymore. I suggest removing it from this document altogether. 

- In the realistic causal graph, there is no need to repeat steps 7-11 N times, because again this is the exercise of seeing whether you recover your simulation parameters if you use the same model specification. Instead, what we will do is repeat the simulation N times, and apply the perturbation at each simulation, and store the average treatment effect. So I would change the steps to the following:

 7. The obtained distributions from step 5 will be used to simulate data for nodes without parents in the specified causal graph.
 8. Data from step 7, alongside with estimated relationships from step 6 will be used to estimate data for the remaining nodes in the causal graph (with the exception of the utility node)
 9. Based on observed availability of alternatives, a function will use the cleaned dataset to simulate an availability matrix for alternatives for the simulated dataset
 10. The specified model from step 3 will be used to simulate choices for the simulated dataset.
 11. The dataset will be used to estimate a choice model and recover the necessary parameters. (These parameters can be stored in a dictionary, if necessary)
 12. One variable will be perturbed to simulate a potential external intervention. This perturbation will result in two modified datasets: one where only the perturbed variable changes (assuming an independent causal graph), and one where descendant children nodes of the perturbed node will also be perturbed following the estimated relationships from step 8.
 13. Causal effects will be estimated using the model estimated in step 11 and the true model specified in step 3 on both datasets. The model from step 3 will be applied only on the dataset resulting from the perturbation of the "true" causal graph, giving us the true causal effect. The model from step 11 will be applied to both datasets from step 12, resulting in our "naive" and "estimated" causal effects. 
 14. Repeat steps 7-13 N times, keeping the same perturbation, and the estimated causal effects will be stored in a dictionary.
 15. Plots will be produced to show the distribution of "estimated" and "true" causal effects

@timothyb0912

Author: I will be adding a brief description of what steps are just procedural vs. what steps help in completing the story of the project.

Documentation, in the form of Jupyter Notebook accompanied by a Markdown file (README.md) detailing the procedure for installing the environment and the content of folders and notebooks. Any limitations to the project will be noted in the jupyter notebooks developed to illustrate the analysis effort. The README.md markdown file will also will point to documentation pages for any packages used in the project.

### 15.2 Is there a plan for sharing the software with others?
- **File upload?**
- **Upload to package repository?**

All project related material including code, documentation, presentation, environment files, and datasets will be uploaded to the project github repository.

### 15.3 Will the project have an associated website? If so is there a plan for its development?

The only "website" for the project will be its github repository.

### 15.4 Is there a plan for producing any written or visual materials that others may wish to view?
- **Final or interim project reports**
- **Tables of key project results**
- **Graphs, diagrams, images from the project, e.g.:**
 - **Plots of system performance**
 - **Model checking plots**

> 

Author: @timothyb0912 As far as I know, the only thing we will be producing will be a presentation to be sent to the conference organizers and uploaded to the repository. Maybe I am overthinking this question, but does it just ask for if we will be producing any presentable visual materials?

Reviewer: Not if, but whether there is (and essentially what is) the plan for producing the visual materials?

What specific plots, tables, diagrams, etc. are we expecting to include?

What parts (e.g. what files / classes / functions) of the architecture / software are going to produce those materials?

Author: Got it.

- **List needed intermediate products (e.g. pseudo-code, unit tests, etc.)**
- **Try to be concise but comprehensive?**

The project is split into three main tasks:

1. The first task is focused on a selection-on-observables problem:

For a causal graph with independent covariates:
 1. Ingest data
 2. Manipulate data as needed, the output of this step will be a cleaned dataset

>  3. Specify a choice model based on the data at hand, the specification should follow required conventions by third party dependencies

Reviewer: ```suggestion
 3. Specify and estimate a choice model based on the data at hand, the specification should follow required conventions by third party dependencies
```

Without that change, steps 10 and 12 don't make sense (as we wouldn't have the "true" parameters that are estimated at step 3).

Author: Right!

 7. The specified model from step 3 will be used to simulate choices for the simulated dataset.
 8. The dataset will be used to estimate a choice model and recover the necessary parameters. These parameters can be stored in a dictionary.
 9. Steps 5-8 will be repeated for N times (say 100) and the parameters from each iteration will be stored in a dictionary entry.
 10. These paramaters will be stored in a dictionary that will be used in a function to plot the original parameters used to simulate the choices (from step 3) along with any recovered parameters from step 9.
 11. One variable will be perturbed to simulate a potential external intervention. In this scenario, no other variables will be changes since all covariates are independent.
 12. Causal effects will be estimated using the models estimated in step 3 and 9.
 13. Plots will be produced to show the distribution of "naive" and "true" causal effects

For a realistic causal graph with confounding:
 1. Ingest data
 2. Manipulate data as needed, the output of this step will be a cleaned dataset

>  3. Specify a choice model based on the data at hand, the specification should follow required conventions by third party dependencies

Reviewer: These steps are only performed once correct? I'm of the impression that you reuse the same cleaned dataset and choice model, and "true parameters" estimated from that specified model and the original data for both the naive and realistic causal graphs right? If so, I would begin with step 4 below and remove the first three steps (or mention that the results of steps 1 - 3 from the naive scenario are reused).

Author: Ah! Agreed, this avoids duplication.

 8. Data from step 7, alongside with estimated relationships from step 6 will be used to estimate data for the remaining nodes in the causal graph (with the exception of the utility node)
 9. Based on observed availability of allternatives, a function will use the cleaned dataset to simulate an availability matrix for alternatives for the simulated dataset
 10. The specified model from step 3 will be used to simulate choices for the simulated dataset.
 11. The dataset will be used to estimate a choice model and recover the necessary parameters. These parameters can be stored in a dictionary.
 12. Steps 7-11 will be repeated for N times (say 100) and the parameters from each iteration will be stored in a dictionary entry.
 13. These paramaters will be stored in a dictionary that will be used in a function to plot the original parameters used to simulate the choices (from step 3) along with any recovered parameters from step 9.
 14. One variable will be perturbed to simulate a potential external intervention. In this scenario, descendant children nodes of the perturbed node will also be perturbed following the estimated relationships from step 8.

>  15. Causal effects will be estimated using the models estimated in step 3 and 121.

Reviewer: ```suggestion
 15. Causal effects will be estimated using the models estimated in step 3 and 12.
```

- **Flyte?**
- **Make?**
- **Other?**  
We don't have any plans for this. please make any suggestions as you see needed/fit.

## 8. Description of experimentation protocols (if relevant):
**If experiments need to be run (e.g. hyper-parameter tuning, model-selection, A/B tests, etc.):**
### 8.1 How will experiments be launched?
N/A

### 8.2 How will experiment reproducibility be ensured?
N/A

### 8.3 How will experiment meta-data (e.g. launch configurations) be stored?
N/A

### 8.4 How will experiment-created data (e.g. results) be stored?
N/A

### 8.5 How will experiment-created data be analyzed?
N/A

### 8.6 How will experiment analyses be prepared for public reporting? 
N/A

## 9. Description of user-interface:
### 9.1 How will users (including myself) interact with the system that is built?
- **Will there be a command line interface?**
- **Will users be editing configuration files, and if so, where will they be stored?**
- **Will there be ability to replay the DAG representing one’s data analysis?**

Users will use a jupyter notebook to interact with the system that is build

> No configuration files will need to be edited. If any functions need to be edited, one could easily make the adjustment

Reviewer: I strongly advise against this.

At the very least, all parameters for the notebook should be defined in a cell at the top (the idea is the notebooks in their final state should at least be runnable via the papermill package).

Not allowing the code for the project to be run outside the notebook is highly limiting if you or someone else wants to reuse a notebook as part of a pipeline elsewhere.

We should really strive to make life as easy as possible for others, especially if it doesn't require us going to a heroic extent (like changing how we organize a notebook). Just my suggestion!

Author: Ah! I got what you mean now. My interpretation of configuration files was project environment configuration files (requirements.txt for example). Sorry for misunderstanding.

I think the config file should contain data relevant to any arguments or parameters (including dataset urls) that are to be used in the project. I will update the answer accordingly.

Reviewer: No worries, I'm learning how to communicate more clearly about these things!

Author: No worries, really appreciate the direction.

## 2. Description of major components of the system and their relations:
### 2.1 What should each component do?

> 1. The goal of showing the need of considering treatment assignment mechanism when estimating causal effects requires show an application of the steps outlined in Brathwaite and Walker (2018a) to a simulated dataset from Brathwaite and Walker (2018b). This application will consitute of the following classes/functions:

Reviewer: ```suggestion
1. Show the need for considering the treatment assignment mechanism when estimating causal effects. This requires an application of the steps outlined in Brathwaite and Walker (2018a) to a simulated dataset from Brathwaite and Walker (2018b). This application will contain the following classes/functions:
```

### 2.2 What is the interface between each component and each other component?
Each component(classes) will be written to be as independent as possible from other components. Functions within classes
will depend on each other.
However, we will write functions that test the output of each of our functions to make sure we get the expected output.

> In general, the interface between each of the project component will be some sort of data: dictionaries storing distribution parameters,

Reviewer: ```suggestion
In general, the interface between each of the project components will be some sort of data: dictionaries storing distribution parameters,
```

> simulated data, estimated model paramters, causal graphs, and arrays of statistics abour estimated model parameters. **this list is not

Reviewer: ```suggestion
simulated data, estimated model parameters, causal graphs, and arrays of statistics about estimated model parameters. **this list is not
```

Summaries of model estimate will be saved as markdown tables

Plots of results will be saved as .png files and pushed to the repo and included in the final presentation

Previously estimated models will be stored in a dictionary that can be pickled and loaded later.

> ### 7.6 How will created data be validated?

Reviewer: Below, I imagine it should be "Ensure all columns have the correct types" as opposed to `tyoes`.

Author: yes. will change.

- **Flyte?**
- **Make?**
- **Other?**  
Developing a project workflow file that shows how all specified requirements of each workflow
step are met. 

## 8. Description of experimentation protocols (if relevant):
**If experiments need to be run (e.g. hyper-parameter tuning, model-selection, A/B tests, etc.):**
### 8.1 How will experiments be launched?

> We can store launch the experiment from a jupyter notebook, and **hopefully** from the command line.

Reviewer: ```suggestion
We can launch the experiment from a jupyter notebook, and **hopefully** from the command line.
```

and parameters stored in a .json/pickle/csv file. the sacred library has a `run` object
that stores all experiment info and configuration.

### 8.4 How will experiment-created data (e.g. results) be stored?
The experiment-created data will be stored in json/pickle/csv files.

### 8.5 How will experiment-created data be analyzed?
Experiment metrics will be analyzed to find statistics about means and variances of parameters
to make sure they fall within the expected ranges.

### 8.6 How will experiment analyses be prepared for public reporting? 
Experiment analyses will be stored in .json/csv files and used to create plots summarizing
results of the experiments.

## 9. Description of user-interface:
### 9.1 How will users (including myself) interact with the system that is built?
- **Will there be a command line interface?**
- **Will users be editing configuration files, and if so, where will they be stored?**
- **Will there be ability to replay the DAG representing one’s data analysis?**

Users will use a jupyter notebook to interact with the system. The `papermill` library
will be used to parametrize a notebook. Parameters of the notebook could also be saved

> in a .json file that is loaded into the notebook. For experiments, the `sacred` library

Reviewer: Smooth. I looked into sacred before but for whatever reason never fell in love with it. I just checked it out again and it looks like it'll be good!

* All the plots will be generated using functions from the source code. Additional plots as needed will be generated
using matplotlib/seaborn. 

>     - The plots generated will show the distribution of the recovered parameters and where the true parameters fall within that distribution. Or where the recovered parameters fall within the confidence intervals of the "true" parameters assumed in simulating the data.

Reviewer: Your second sentence sounds hard since the data-generating parameters have no confidence interval around them. One only gets confidence intervals around estimated parameters.

For the second sentence, I would instead show the proportion of estimates where the true parameters fall within some percentile confidence interval (e.g. 95th percentile) around the estimate. This would essentially just end up being a bar plot with two bars or a table.

Author: I guess what I meant by "true" parameters is the parameters used to simulate the data. 
Making the change now.

* All the plots will be generated using functions from the source code. Additional plots as needed will be generated
using matplotlib/seaborn. 

>     - The plots generated will show the distribution of the recovered parameters and where the true parameters fall within that distribution.

Reviewer: ```suggestion
    - The plots generated will show the distribution of the estimated causal effects and where the true causal effects fall within that distribution.
```
